File: master_of_plow_CPP_SDK/src-gen/nevonex/can_agmo_mtlt305/CAN_AGMO_MTLT305.cpp
/*
 * nevonex/can_agmo_mtlt305/CAN_AGMO_MTLT305.cpp
 * Copyright (c) Robert Bosch GmbH. All rights reserved.
 */

#include "CAN_AGMO_MTLT305.hpp"
#include <ecore/EObject.hpp>
#include <ecore/EClass.hpp>
#include "nevonex/can_agmo_mtlt305/Can_agmo_mtlt305Package.hpp"
#include <ecorecpp/mapping.hpp>
#include <nevonex-fcal-platform/log/Logger.hpp>

/*PROTECTED REGION ID(CAN_AGMO_MTLT305.cpp) START*/
// Please, enable the protected region if you add manually written code.
// To do this, add the keyword ENABLED before START.
/*PROTECTED REGION END*/

using namespace ::nevonex::can_agmo_mtlt305;
using namespace ::nevonex::log;

// Default constructor
CAN_AGMO_MTLT305::CAN_AGMO_MTLT305() : m_aceinna_Accel(0), m_aceinna_AngleRate(0), m_aceinna_Angles(0)
{
    // Machine does NOT call _initialize() — Provider manages initialization

    m_interfaceDetailMap =
    {
    { 9440, ::nevonex::types::InterfaceDetails("fek/9440") },
    { 9441, ::nevonex::types::InterfaceDetails("fek/9441") },
    { 9442, ::nevonex::types::InterfaceDetails("fek/9442") } };

    /*PROTECTED REGION ID(CAN_AGMO_MTLT305__CAN_AGMO_MTLT305) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

CAN_AGMO_MTLT305::~CAN_AGMO_MTLT305()
{
    /*PROTECTED REGION ID(CAN_AGMO_MTLT305__Destructor) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

// References
::nevonex::fcal::ACEINNA_ACCEL_T_R_ptr CAN_AGMO_MTLT305::getAceinna_Accel() const
{
    return m_aceinna_Accel;
}

void CAN_AGMO_MTLT305::setAceinna_Accel(::nevonex::fcal::ACEINNA_ACCEL_T_R_ptr _aceinna_Accel)
{
    ::nevonex::fcal::ACEINNA_ACCEL_T_R_ptr _old_aceinna_Accel = m_aceinna_Accel;
    m_aceinna_Accel = _aceinna_Accel;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_MTLT305 property \"aceinna_Accel\" has changed.";
    notifyPropertyChange("aceinna_Accel", _old_aceinna_Accel, m_aceinna_Accel);
}

::nevonex::fcal::ACEINNA_ANGLERATE_T_R_ptr CAN_AGMO_MTLT305::getAceinna_AngleRate() const
{
    return m_aceinna_AngleRate;
}

void CAN_AGMO_MTLT305::setAceinna_AngleRate(::nevonex::fcal::ACEINNA_ANGLERATE_T_R_ptr _aceinna_AngleRate)
{
    ::nevonex::fcal::ACEINNA_ANGLERATE_T_R_ptr _old_aceinna_AngleRate = m_aceinna_AngleRate;
    m_aceinna_AngleRate = _aceinna_AngleRate;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_MTLT305 property \"aceinna_AngleRate\" has changed.";
    notifyPropertyChange("aceinna_AngleRate", _old_aceinna_AngleRate, m_aceinna_AngleRate);
}

::nevonex::fcal::ACEINNA_ANGLES_T_R_ptr CAN_AGMO_MTLT305::getAceinna_Angles() const
{
    return m_aceinna_Angles;
}

void CAN_AGMO_MTLT305::setAceinna_Angles(::nevonex::fcal::ACEINNA_ANGLES_T_R_ptr _aceinna_Angles)
{
    ::nevonex::fcal::ACEINNA_ANGLES_T_R_ptr _old_aceinna_Angles = m_aceinna_Angles;
    m_aceinna_Angles = _aceinna_Angles;
    SDK_FCAL_LOG(SeverityLevel::debug) << "CAN_AGMO_MTLT305 property \"aceinna_Angles\" has changed.";
    notifyPropertyChange("aceinna_Angles", _old_aceinna_Angles, m_aceinna_Angles);
}

// FCAL timestamp/valid
::ecore::ELong CAN_AGMO_MTLT305::getAceinna_Accel_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9440);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean CAN_AGMO_MTLT305::isAceinna_Accel_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9440);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

::ecore::ELong CAN_AGMO_MTLT305::getAceinna_AngleRate_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9441);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean CAN_AGMO_MTLT305::isAceinna_AngleRate_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9441);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

::ecore::ELong CAN_AGMO_MTLT305::getAceinna_Angles_Timestamp()
{
    auto _detail = m_interfaceDetailMap.find(9442);
    if (_detail == m_interfaceDetailMap.end())
        return -1;
    return _detail->second.getTimestamp();
}

::ecore::EBoolean CAN_AGMO_MTLT305::isAceinna_Angles_Valid()
{
    auto _detail = m_interfaceDetailMap.find(9442);
    if (_detail == m_interfaceDetailMap.end())
        return false;
    return _detail->second.isValid();
}

::ecore::EBoolean CAN_AGMO_MTLT305::updateInterfaceDetail(int _topicKey, ::ecore::EBoolean _valid, ::ecore::ELong _timestamp)
{
    auto _detail = m_interfaceDetailMap.find(_topicKey);
    if (_detail == m_interfaceDetailMap.end())
    {
        SDK_FCAL_LOG(SeverityLevel::warning) << _topicKey
                << " interface details are not present in the interface detail map[CAN_AGMO_MTLT305].";
        return false;
    }

    if (_detail->second.getTimestamp() > _timestamp)
    {
        SDK_FCAL_LOG(SeverityLevel::warning)
                << "Old or duplicate message received. Old TS : "
                << _detail->second.getTimestamp() << " new TS : " << _timestamp
                << " feature: CAN_AGMO_MTLT305 topic:" << _topicKey;
        return false;
    }

    _detail->second.setValid(_valid, _timestamp);
    return true;
}

// PropertyChange delegation (diamond inheritance requires explicit override)
void CAN_AGMO_MTLT305::addPropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::addPropertyChangeListener(_listener);
}

void CAN_AGMO_MTLT305::removePropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::removePropertyChangeListener(_listener);
}

void CAN_AGMO_MTLT305::notifyPropertyChange(::ecore::EString const &_name,
        ::ecore::EJavaObject const &_oldValue,
        ::ecore::EJavaObject const &_newValue)
{
    using namespace ::nevonex::types;
    PropertyChange::notifyPropertyChange(_name, _oldValue, _newValue);
}

File: com.bosch.fsp.master_of_plow.gen/src/com/bosch/nevonex/can_agmo_mtlt305/impl/CAN_AGMO_MTLT305.java
/**
Copyright (c) Robert Bosch GmbH. All rights reserved.
*/
package com.bosch.nevonex.can_agmo_mtlt305.impl;

import com.bosch.fsp.logger.FCALLogs;
import com.bosch.fsp.logger.LoggerConstants;

import com.bosch.fsp.platform.InterfaceDetails;

import com.bosch.fsp.runtime.feature.IMachine;

import com.bosch.nevonex.common.ITopicObject;
import com.bosch.nevonex.common.impl.TopicObject;

import com.bosch.nevonex.fcal.IACEINNA_ACCEL_T_R;

import com.bosch.nevonex.fcal.IACEINNA_ANGLERATE_T_R;

import com.bosch.nevonex.fcal.IACEINNA_ANGLES_T_R;

import com.bosch.nevonex.can_agmo_mtlt305.ICAN_AGMO_MTLT305;

import com.bosch.nevonex.types.IPropertyChange;
import com.bosch.nevonex.types.impl.TypesPackage;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import java.lang.reflect.InvocationTargetException;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.eclipse.emf.common.util.BasicEList;
import org.eclipse.emf.common.util.EList;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.util.EDataTypeUniqueEList;

/**
 * An implementation of the model object '<em><b>CAN_AGMO_MTLT305</b></em>'.
 * @generated
 */
public class CAN_AGMO_MTLT305 extends TopicObject implements ICAN_AGMO_MTLT305 {
	protected EList<PropertyChangeListener> listeners;
	protected IACEINNA_ACCEL_T_R aceinna_Accel;
	protected IACEINNA_ANGLERATE_T_R aceinna_AngleRate;
	protected IACEINNA_ANGLES_T_R aceinna_Angles;

	/**
	 * @generated
	 */
	protected CAN_AGMO_MTLT305() {
		super();
		initInterfaceDetails();
	}

	private void initInterfaceDetails() {
		interfaceDetailMap.put("fek/9440", new InterfaceDetails("fek/9440"));
		interfaceDetailMap.put("fek/9441", new InterfaceDetails("fek/9441"));
		interfaceDetailMap.put("fek/9442", new InterfaceDetails("fek/9442"));
	}

	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return Can_agmo_mtlt305Package.Literals.CAN_AGMO_MTLT305;
	}

	/**
	 * @generated
	 */
	public IACEINNA_ACCEL_T_R getAceinna_Accel() {
		if (aceinna_Accel != null && aceinna_Accel.eIsProxy()) {
			InternalEObject oldAceinna_Accel = (InternalEObject) aceinna_Accel;
			aceinna_Accel = (IACEINNA_ACCEL_T_R) eResolveProxy(oldAceinna_Accel);
			if (aceinna_Accel != oldAceinna_Accel) {
			}
		}
		return aceinna_Accel;
	}

	/**
	 * @generated
	 */
	public IACEINNA_ACCEL_T_R basicGetAceinna_Accel() {
		return aceinna_Accel;
	}

	/**
	 * @generated
	 */
	public void setAceinna_Accel(IACEINNA_ACCEL_T_R newAceinna_Accel) {
		IACEINNA_ACCEL_T_R oldAceinna_Accel = aceinna_Accel;
		aceinna_Accel = newAceinna_Accel;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " CAN_AGMO_MTLT305 property \"aceinna_Accel\" changed from  "
						+ oldAceinna_Accel + " to " + newAceinna_Accel);
		notifyPropertyChange("aceinna_Accel", oldAceinna_Accel, newAceinna_Accel);
	}

	/**
	 * @generated
	 */
	public long getAceinna_Accel_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9440");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isAceinna_Accel_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9440");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public IACEINNA_ANGLERATE_T_R getAceinna_AngleRate() {
		if (aceinna_AngleRate != null && aceinna_AngleRate.eIsProxy()) {
			InternalEObject oldAceinna_AngleRate = (InternalEObject) aceinna_AngleRate;
			aceinna_AngleRate = (IACEINNA_ANGLERATE_T_R) eResolveProxy(oldAceinna_AngleRate);
			if (aceinna_AngleRate != oldAceinna_AngleRate) {
			}
		}
		return aceinna_AngleRate;
	}

	/**
	 * @generated
	 */
	public IACEINNA_ANGLERATE_T_R basicGetAceinna_AngleRate() {
		return aceinna_AngleRate;
	}

	/**
	 * @generated
	 */
	public void setAceinna_AngleRate(IACEINNA_ANGLERATE_T_R newAceinna_AngleRate) {
		IACEINNA_ANGLERATE_T_R oldAceinna_AngleRate = aceinna_AngleRate;
		aceinna_AngleRate = newAceinna_AngleRate;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " CAN_AGMO_MTLT305 property \"aceinna_AngleRate\" changed from  "
						+ oldAceinna_AngleRate + " to " + newAceinna_AngleRate);
		notifyPropertyChange("aceinna_AngleRate", oldAceinna_AngleRate, newAceinna_AngleRate);
	}

	/**
	 * @generated
	 */
	public long getAceinna_AngleRate_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9441");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isAceinna_AngleRate_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9441");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public IACEINNA_ANGLES_T_R getAceinna_Angles() {
		if (aceinna_Angles != null && aceinna_Angles.eIsProxy()) {
			InternalEObject oldAceinna_Angles = (InternalEObject) aceinna_Angles;
			aceinna_Angles = (IACEINNA_ANGLES_T_R) eResolveProxy(oldAceinna_Angles);
			if (aceinna_Angles != oldAceinna_Angles) {
			}
		}
		return aceinna_Angles;
	}

	/**
	 * @generated
	 */
	public IACEINNA_ANGLES_T_R basicGetAceinna_Angles() {
		return aceinna_Angles;
	}

	/**
	 * @generated
	 */
	public void setAceinna_Angles(IACEINNA_ANGLES_T_R newAceinna_Angles) {
		IACEINNA_ANGLES_T_R oldAceinna_Angles = aceinna_Angles;
		aceinna_Angles = newAceinna_Angles;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " CAN_AGMO_MTLT305 property \"aceinna_Angles\" changed from  "
						+ oldAceinna_Angles + " to " + newAceinna_Angles);
		notifyPropertyChange("aceinna_Angles", oldAceinna_Angles, newAceinna_Angles);
	}

	/**
	 * @generated
	 */
	public long getAceinna_Angles_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9442");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isAceinna_Angles_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("fek/9442");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public void addPropertyChangeListener(PropertyChangeListener listener) {
		if (listeners == null) {
			listeners = new BasicEList<>();
		}
		listeners.add(listener);
	}

	/**
	 * @generated
	 */
	public void removePropertyChangeListener(PropertyChangeListener listener) {
		if (listeners != null) {
			listeners.remove(listener);
		}
	}

	/**
	 * @generated
	 */
	public void notifyPropertyChange(String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(this, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ACCEL:
			if (resolve)
				return getAceinna_Accel();
			return basicGetAceinna_Accel();
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLE_RATE:
			if (resolve)
				return getAceinna_AngleRate();
			return basicGetAceinna_AngleRate();
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLES:
			if (resolve)
				return getAceinna_Angles();
			return basicGetAceinna_Angles();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * @generated
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ACCEL:
			setAceinna_Accel((IACEINNA_ACCEL_T_R) newValue);
			return;
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLE_RATE:
			setAceinna_AngleRate((IACEINNA_ANGLERATE_T_R) newValue);
			return;
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLES:
			setAceinna_Angles((IACEINNA_ANGLES_T_R) newValue);
			return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ACCEL:
			setAceinna_Accel((IACEINNA_ACCEL_T_R) null);
			return;
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLE_RATE:
			setAceinna_AngleRate((IACEINNA_ANGLERATE_T_R) null);
			return;
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLES:
			setAceinna_Angles((IACEINNA_ANGLES_T_R) null);
			return;
		}
		super.eUnset(featureID);
	}

	/**
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ACCEL:
			return aceinna_Accel != null;
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLE_RATE:
			return aceinna_AngleRate != null;
		case Can_agmo_mtlt305Package.CAN_AGMO_MTLT305__ACEINNA_ANGLES:
			return aceinna_Angles != null;
		}
		return super.eIsSet(featureID);
	}
} //CAN_AGMO_MTLT305
